src/main/scala/dpla/ingestion3/MappingEntry.scala

```diff
@@ -71,36 +78,34 @@ object MappingEntry {
     }
 
     // Run the mapping over the Dataframe
-    val mappingResults = harvestedRecords.select("document").map(
-      record => {
-        extractorClass.getConstructor(classOf[String])
-          .newInstance(record.getAs[String]("document"))
-          .build
-      }
-    )
-
-    // TODO there is probably a much cleaner/better way of writing this.
-    val mappingSuccess = mappingResults
-      .filter(r => r.isInstanceOf[DplaMapData])
-      .map(r2 => r2.asInstanceOf[DplaMapData])
-
-    // TODO ditto ^^ This converts the Dataset to an Array[String], This is an easy way out for now.
-    val failures = mappingResults
-      .filter(r => r.isInstanceOf[DplaMapError])
-      .map(r2 => r2.asInstanceOf[DplaMapError]).map(f => f.errorMessage).collect()
+    val documents = harvestedRecords.select("document").as[String]
+    val mappingResults = documents.map(document => map(extractorClass, document))(tupleRowStringEncoder)
 
     // Delete the output location if it exists
     Utils.deleteRecursively(new File(dataOut))
 
-    // Save successfully mapped records out to Avro file
-    mappingSuccess.toDF("document").write
-      .format("com.databricks.spark.avro")
-      .save(dataOut)
+
+    val successResults = mappingResults
+          .filter(_._1 != null)
+          .map(row => row._1)(dplaMapDataRowEncoder)
+
+    val failures = mappingResults.filter(_._2 != null).map(_._2).collect()
+
+    successResults.toDF().write.avro(dataOut)
+
 
     // Summarize results
-    mappingSummary(harvestedRecords.count(), mappingSuccess.count(),failures, dataOut, shortName)
+    mappingSummary(harvestedRecords.count(), successResults.count(), failures, dataOut, shortName)
+
+    spark.stop()
   }
 
+  private def map(extractorClass: Class[_ <: Extractor], document: String): (Row, String) =
+    extractorClass.getConstructor(classOf[String]).newInstance(document).build() match {
+      case Success(dplaMapData) => (RowConverter.toRow(dplaMapData, model.sparkSchema), null)
+      case Failure(exception) => (null, exception.getMessage)
```

Comment:
What is the reason for using `null` in the two lines above, as opposed to `Option`?

Reply:
I could be wrong, but I had trouble dealing with Option in Dataframes and Datasets. They just aren't meant for that. I don't know that I tried it in this particular code, so I'll try it quickly to make sure.

Reply:
This seems to back up what I have found so far, that it's non-trivial: https://stackoverflow.com/questions/36449368/how-to-create-encoder-for-option-type-constructor-e-g-optionint

Reply:
Ah, that makes sense.

Reply:
It does, only because they need to give us more options for encoding... Options.